framework/src/bundle/CoreBundleContext.cpp

```diff
@@ -169,20 +169,32 @@ void CoreBundleContext::Uninit1()
   listeners.Clear();
   resolver.Clear();
 
+  // Do not hold the bundleThreads lock while calling
+  // BundleThread::Quit() because this will join the bundle
+  // thread which may need to acquire the bundleThreads lock
+  // itself.
+  //
+  // At this point, all bundles are stopped and all
+  // bundle context instances invalidated. So no new bundle
+  // threads can be created and it is sufficient to get
+  // the current list once and not check for new bundle
+  // threads again.
+  std::list<std::shared_ptr<BundleThread>> threads;
+  bundleThreads.Lock(), std::swap(threads, bundleThreads.value);
```

Comment:
Oh my, you are right. This is again not safe. `BundleThread::Join()` will not help much as far as I can see (it does not set the `doRun` flag to false compared to `Quit()` but then executes the same logic) and will just wait longer (until the keep-alive is exceeded). However, there is a more fundamental problem: We might always miss out some bundle threads when we copy the `bundleThreads.value` list, namely the currently busy ones. I will give it some more thought.